File: output/ChatGPT/avatar/Java/C++/atcoder_ABC126_E.cpp
#include <iostream>
#include <vector>
using namespace std;

class UnionFindTree {
    private:
        int n;
        vector<int> par;
        vector<int> rank;

    public:
        UnionFindTree(int n) {
            this->n = n;
            par.resize(n);
            rank.resize(n);
            for (int i = 0; i < n; i++) {
                par[i] = i;
                rank[i] = 0;
            }
        }

        int find(int x) {
            if (par[x] == x) {
                return x;
            } else {
                return par[x] = find(par[x]);
            }
        }

        void unite(int x, int y) {
            x = find(x);
            y = find(y);
            if (x == y) return;
            if (rank[x] < rank[y]) par[x] = y;
            else {
                par[y] = x;
                if (rank[x] == rank[y]) rank[x]++;
            }
        }

        bool same(int x, int y) {
            return find(x) == find(y);
        }
};

void doIt() {
    int n;
    cin >> n;
    UnionFindTree utf(n);
    int m;
    cin >> m;
    for (int i = 0; i < m; i++) {
        int x, y, z;
        cin >> x >> y >> z;
        utf.unite(x-1, y-1);
    }
    int cnt = 0;
    for (int i = 0; i < n; i++) {
        if (utf.find(i) == i) {
            cnt++;
        }
    }
    cout << cnt << endl;
}

int main() {
    doIt();
    return 0;
}

File: output/ChatGPT/avatar/Java/Python/atcoder_ABC126_E.py
class UnionFindTree:
    def __init__(self, n):
        self.n = n
        self.par = [i for i in range(n)]
        self.rank = [0 for _ in range(n)]

    def find(self, x):
        if self.par[x] == x:
            return x
        else:
            self.par[x] = self.find(self.par[x])
            return self.par[x]

    def unite(self, x, y):
        x = self.find(x)
        y = self.find(y)
        if x == y:
            return
        if self.rank[x] < self.rank[y]:
            self.par[x] = y
        else:
            self.par[y] = x
            if self.rank[x] == self.rank[y]:
                self.rank[x] += 1

    def same(self, x, y):
        return self.find(x) == self.find(y)


def doIt():
    n = int(input())
    utf = UnionFindTree(n)
    m = int(input())
    for i in range(m):
        x, y, z = map(int, input().split())
        utf.unite(x - 1, y - 1)
    cnt = 0
    for i in range(n):
        if utf.par[i] == i:
            cnt += 1
    print(cnt)


doIt()